numpy/core/src/umath/ufunc_object.c

```diff
@@ -2336,10 +2408,11 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
         /* Except for when it belongs to this output */
         if (i >= nin) {
             int dim_offset = ufunc->core_offsets[i];
-            int num_dims = ufunc->core_num_dims[i];
+            int num_dims = core_num_dims[i];
+
             /* Fill in 'iter_shape' and 'op_axes' for this output */
             for (idim = 0; idim < num_dims; ++idim) {
-                iter_shape[j] = core_dim_sizes[
+                iter_shape[j] = keepdims ? 1 : core_dim_sizes[
                                         ufunc->core_dim_ixs[dim_offset + idim]];
                 op_axes_arrays[i][j] = REMAP_AXIS(i, n + idim);
                 ++j;
```

Comment:
Unrelated to this PR: I'm struggling to understand what `j` is, and why its state is kept between outputs

Reply:
I agree, I turn from thinking it clear to having no clue almost every time I look at this...  Mostly because I only partially understand how the iterator works. But this would be a good place to comment it more heavily. For another PR...